src/xdsl/dialects/builtin.py

```diff
@@ -204,3 +251,62 @@ class ModuleOp(Operation):
     @property
     def ops(self) -> List[Operation]:
         return self.regions[0].blocks[0].ops
+
+
+@overload
+def func(name: str, type: Attribute, return_type: Attribute,
+         f: Callable[[BlockArgument], List[Operation]]) -> Operation:
+    ...
+
+
+@overload
+def func(
+        name: str, types: List[Attribute], return_type: Attribute,
+        f: Callable[[BlockArgument, BlockArgument],
+                    List[Operation]]) -> Operation:
+    ...
+
+
+@overload
+def func(
+    name: str, types: List[Attribute], return_type: Attribute,
+    f: Callable[[BlockArgument, BlockArgument, BlockArgument], List[Operation]]
+) -> Operation:
+    ...
+
+
+def func(name, input_types, return_types, f) -> Operation:
+    type_attr = FunctionType.build(input_types, return_types)
+    op = FuncOp.create(
+        [], [],
+        attributes={
+            "sym_name": StringAttr.build(name),
+            "type": type_attr,
+            "sym_visibility": StringAttr.from_str("private")
+        },
+        regions=[Region([block(input_types, f)])])
+    return op
+
+
+# This function is easier to use while parsing, but makes the
+# inline definitions complicated
+def func2(name, input_types, return_types, region: Region) -> Operation:
```

Comment:
I think that we should try to eliminate either `func` or `func2` and force the users to use `FuncOp.create` themselves for one of the cases. 

Removing `func2` might be better, as it uses `FuncOp.create` in a more or less straight forward manner.

Reply:
I'll take care of operation builders in a next PR! this should simplify this case.